File: leetcode-js/problems/median-of-two-sorted-arrays.js
// https://leetcode.com/problems/median-of-two-sorted-arrays

import assert from "assert";

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
function findMedianSortedArrays(nums1, nums2) {
  if (nums1.length > nums2.length) {
    [nums1, nums2] = [nums2, nums1];
  }

  let iMin = 0;
  let iMax = nums1.length - 1;

  let i, j;

  const half = Math.floor((nums1.length + nums2.length) / 2);

  while (true) {
    i = Math.floor((iMin + iMax) / 2);
    j = half - i - 2;

    if (i >= 0 && i < nums1.length && j + 1 < nums2.length && nums1[i] > nums2[j + 1]) {
      iMax = i - 1;
    } else if (j >= 0 && i + 1 < nums1.length && j < nums2.length && nums1[i + 1] < nums2[j]) {
      iMin = i + 1;
    } else {
      break;
    }
  }

  console.log(i, iMin, iMax);

  if ((nums1.length + nums2.length) % 2 === 0) {
    return average(max(nums1[i], nums2[j]), min(nums1[i + 1], nums2[j + 1]));
  }

  return min(nums1[i + 1], nums2[j + 1]);
}

function max(x, y) {
  if (x == null) return y;
  if (y == null) return x;
  return Math.max(x, y);
}

function min(x, y) {
  if (x == null) return y;
  if (y == null) return x;
  return Math.min(x, y);
}

function average(x, y) {
  if (x == null) return y;
  if (y == null) return x;
  return (x + y) / 2;
}

function findMedian(nums) {
  return (nums[Math.floor((nums.length - 1) / 2)] + nums[Math.ceil((nums.length - 1) / 2)]) / 2;
}

// prettier-ignore
const arrayPairs = [
  [[1, 3, 4, 6, 8, 10, 11, 20], [2, 5, 7, 12, 13, 14, 21]],
  [[1, 5, 6, 7, 10],            [3, 4]],
  [[5, 6, 7, 10, 11],           [1, 2, 3]],
  [[5],                         [1, 2, 3]],
  [[6],                         [1, 2]],
  [[5, 6, 7, 8],                [1]],
  [[6, 7, 8, 9],                [2, 3]],
  [[2],                         []]
];

for (const [nums1, nums2] of arrayPairs) {
  const nums = nums1.concat(nums2);
  nums.sort((a, b) => a - b);
  const expectedResult = findMedian(nums);
  const result = findMedianSortedArrays(nums1, nums2);
  assert(Math.abs(expectedResult - result) < 0.001, `Expected ${expectedResult}, instead got ${result}.\n${JSON.stringify(nums1)} ${JSON.stringify(nums2)}`);
}

console.log(
  findMedianSortedArrays([1, 3, 4, 5, 9, 14, 15, 20], [2, 6, 16, 17])
)

File: leetcode-rs/src/_median_of_two_sorted_arrays.rs
// https://leetcode.com/problems/median-of-two-sorted-arrays

pub struct Solution;

impl Solution {
  pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
    let mut i = 0;
    let mut j = 0;
    let mut x = 0f64;
    let mut y = 0f64;

    while i + j <= (nums1.len() + nums2.len()) / 2 {
      if i < nums1.len() && (j >= nums2.len() || nums1[i] <= nums2[j]) {
        x = y;
        y = nums1[i] as f64;
        i += 1;
      } else if j < nums2.len() && (i >= nums1.len() || nums1[i] >= nums2[j]) {
        x = y;
        y = nums2[j] as f64;
        j += 1;
      } else {
        panic!("Impossible!");
      }
    }

    if (nums1.len() + nums2.len()) % 2 == 0 {
      (x + y) / 2f64
    } else {
      y
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  fn assert_approx(nums1: Vec<i32>, nums2: Vec<i32>, expected_result: f64) {
    let result = Solution::find_median_sorted_arrays(nums1, nums2);
    assert!(
      (result - expected_result).abs() < f64::EPSILON,
      "Expected {expected_result}, instead got {result}."
    );
  }

  #[test]
  pub fn median_test() {
    assert_approx(vec![1, 3, 4, 7, 9, 10, 12], vec![5, 6, 13, 14], 7f64);
    assert_approx(vec![1, 3, 4, 5, 9, 14, 15, 20], vec![2, 6, 16, 17], 7.5f64);
    assert_approx(vec![1, 3], vec![2], 2f64);
    assert_approx(vec![1, 2], vec![3, 4], 2.5f64);
    assert_approx(vec![1, 3], vec![2, 4], 2.5f64);
  }
}
